placement/service/placementservice.go

```diff
@@ -233,39 +222,112 @@ func (ps placementService) findReplaceHost(s placement.Snapshot, candidateHosts
+				conflicts++
+			}
+		}
+		for _, host := range hostsInRack {
+			hosts = append(hosts, sortableThing{thing: host, value: conflicts})
+		}
+	}
+
+	groups := groupHostsByConflict(hosts, ps.options.LooseRackCheck())
+	if len(groups) == 0 {
+		return nil, errNoValidHost
 	}
 
-	return ps.findHostWithRackCheck(s, candidateRackHostMap, leaving)
+	result, leftWeight := fillWeight(groups, int(leaving.Host().Weight()))
+
+	if leftWeight > 0 {
+		return nil, fmt.Errorf("could not find enough host to replace %s, %v weight could not be replaced",
+			leaving.Host().String(), leftWeight)
+	}
+	return result, nil
 }
 
-func (ps placementService) findHostWithRackCheck(p placement.Snapshot, rackHostMap map[string][]placement.Host, leaving placement.HostShards) (placement.Host, error) {
-	// otherwise sort the candidate hosts by the number of conflicts
-	ph := algo.NewPlacementHelper(ps.options, p)
+func groupHostsByConflict(hostsSortedByConflicts []sortableThing, allowConflict bool) [][]placement.Host {
+	sort.Sort(sortableThings(hostsSortedByConflicts))
+	var groups [][]placement.Host
+	lastSeenConflict := -1
+	for _, host := range hostsSortedByConflicts {
+		if !allowConflict && host.value > 0 {
+			break
+		}
+		if host.value > lastSeenConflict {
+			lastSeenConflict = host.value
+			groups = append(groups, []placement.Host{})
+		}
+		if lastSeenConflict == host.value {
+			groups[len(groups)-1] = append(groups[len(groups)-1], host.thing.(placement.Host))
+		}
+	}
+	return groups
+}
 
-	rackLens := make(rackLens, 0, len(rackHostMap))
-	for rack := range rackHostMap {
-		rackConflicts := 0
-		for _, shard := range leaving.Shards() {
-			if !ph.HasNoRackConflict(shard, leaving, rack) {
-				rackConflicts++
-			}
+func fillWeight(groups [][]placement.Host, targetWeight int) ([]placement.Host, int) {
+	var (
+		result       []placement.Host
+		hostsInGroup []placement.Host
+	)
+	for _, group := range groups {
+		hostsInGroup, targetWeight = knapsack(group, targetWeight)
+		result = append(result, hostsInGroup...)
+		if targetWeight <= 0 {
+			break
 		}
 
-		rackLens = append(rackLens, rackLen{rack: rack, len: rackConflicts})
 	}
-	sort.Sort(rackLens)
-	if !ps.options.LooseRackCheck() {
-		rackLens = filterConflictRacks(rackLens)
+	return result, targetWeight
+}
+
+func knapsack(hosts []placement.Host, targetWeight int) ([]placement.Host, int) {
+	totalWeight := 0
+	for _, host := range hosts {
+		totalWeight += int(host.Weight())
+	}
+	if totalWeight <= targetWeight {
+		return hosts[:], targetWeight - totalWeight
+	}
+	// totalWeight > targetWeight, there is a combination of hosts to meet targetWeight for sure
+	// we do dp until totalWeight rather than targetWeight here because we need to cover the targetWeight
+	// which is a little bit different than the knapsack problem that goes
+	weights := make([]int, totalWeight+1)
+	combination := make([][]placement.Host, totalWeight+1)
+
+	for _, host := range hosts {
```

Comment:
Can we add some comments here to explain the DP relationship and that we are preferring assignments that have fewer hosts in the assignment?